Applications/Piano/AudioEngine.h

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2019-2020, William McPherson <willmcpherson2@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "Music.h"
+#include <AK/FixedArray.h>
+#include <AK/Queue.h>
+
+class AudioEngine {
+public:
+    explicit AudioEngine();
+    ~AudioEngine();
+
+    Sample* buffer() const { return m_front_buffer_ptr; }
```

Comment:
This should return `const Sample*` since it's not expected that clients of the function will modify the `Sample` array.

Also, this API is bad since callers are expected to just know that the buffer size is `sample_count`.
How about we make the buffers be something like `FixedArray<Sample>` instead.
Then this function could return a `const FixedArray<Sample>&` and the caller would be able to learn the array size.

Reply:
Do you mean make all our buffers `FixedArray`s? Is the allocation worse?

Reply:
Yes, we should use some sort of smart array type for the buffers. `FixedArray` seems the most appropriate since we don't need to dynamically shrink/grow it after construction (which is what `Vector` is good at.)

The reason using a plain buffer allocation is bad is because then everyone who uses that buffer is expected to magically know how large it is.